<h1 style="color: green">Quantitative Analysis of Stock Market</h1>

In [2]:
""" 
Pandas is an open-source data manipulation and analysis library
Plotly is an interactive data visualization library that allows you to create interactive and visually appealing plots and dashboard
Matplotlib is a popular 2D plotting library for Python that provides a wide variety of static, animated, and interactive plots
nbformat is a Python library that provides tools for working with Jupyter notebooks """

%pip install pandas   
%pip install plotly     
%pip install nbformat    

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [3]:
import pandas as pd

import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.io as pio
pio.templates.default = "plotly_white"

""" 
plotly.express - It simplifies the process of creating complex visualizations with minimal code.
plotly.graph_objects - Plotly Graph Objects provides a lower-level interface for creating more customized and detailed visualizations compared to Plotly Express
plotly.subplots - This function allows you to create subplots within a single figure in Plotly
plotly.io - The plotly.io module in Plotly is used for working with input and output operations, including reading and writing Plotly figures, configuring plot settings, and handling templates.
"""

import matplotlib.pyplot as plt

C:\Users\B Sai Soumath\AppData\Local\Temp\ipykernel_12828\2598418745.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [4]:
stocks_data = pd.read_csv("stocks.csv")
stocks_data

Ticker        Date        Open        High         Low       Close  \
0     AAPL  2023-02-07  150.639999  155.229996  150.639999  154.649994   
1     AAPL  2023-02-08  153.880005  154.580002  151.169998  151.919998   
2     AAPL  2023-02-09  153.779999  154.330002  150.419998  150.869995   
3     AAPL  2023-02-10  149.460007  151.339996  149.220001  151.009995   
4     AAPL  2023-02-13  150.949997  154.259995  150.919998  153.850006   
..     ...         ...         ...         ...         ...         ...   
243   GOOG  2023-05-01  107.720001  108.680000  107.500000  107.709999   
244   GOOG  2023-05-02  107.660004  107.730003  104.500000  105.980003   
245   GOOG  2023-05-03  106.220001  108.129997  105.620003  106.120003   
246   GOOG  2023-05-04  106.160004  106.300003  104.699997  105.209999   
247   GOOG  2023-05-05  105.320000  106.440002  104.738998  106.214996   

      Adj Close    Volume  
0    154.414230  83322600  
1    151.688400  64120100  
2    150.639999  56007100  
3    151.009995  57450700  
4    153.850006  62199000  
..          ...       ...  
243  107.709999  20926300  
244  105.980003  20343100  
245  106.120003  17116300  
246  105.209999  19780600  
247  106.214996  20705300  

[248 rows x 8 columns]

## Columns of Stock Market Data

- Ticker: The stock ticker symbol.
<dl>
<dd>"AAPL" - Apple Inc.</dd>
<dd>"GOOG" - Google/Alphabet Inc.</dd>
<dd>"MSFT" - Microsoft Corporation</dd>
<dd>"NFLX" - Netflix Inc.</dd>
</dl>

- Date: The trading date.
- Open: The opening price of the stock for the day.
- High: The highest price of the stock during the day.
- Low: The lowest price of the stock during the day.
- Close: The closing price of the stock for the day.
- Adj Close: The adjusted closing price, which accounts for all corporate actions such as dividends, stock splits, etc.
- Volume: The number of shares traded during the day.

In [5]:
stocks_data.head()

Ticker        Date        Open        High         Low       Close  \
0   AAPL  2023-02-07  150.639999  155.229996  150.639999  154.649994   
1   AAPL  2023-02-08  153.880005  154.580002  151.169998  151.919998   
2   AAPL  2023-02-09  153.779999  154.330002  150.419998  150.869995   
3   AAPL  2023-02-10  149.460007  151.339996  149.220001  151.009995   
4   AAPL  2023-02-13  150.949997  154.259995  150.919998  153.850006   

    Adj Close    Volume  
0  154.414230  83322600  
1  151.688400  64120100  
2  150.639999  56007100  
3  151.009995  57450700  
4  153.850006  62199000

In [13]:
stocks_data.tail()

Ticker        Date        Open        High         Low       Close  \
243   GOOG  2023-05-01  107.720001  108.680000  107.500000  107.709999   
244   GOOG  2023-05-02  107.660004  107.730003  104.500000  105.980003   
245   GOOG  2023-05-03  106.220001  108.129997  105.620003  106.120003   
246   GOOG  2023-05-04  106.160004  106.300003  104.699997  105.209999   
247   GOOG  2023-05-05  105.320000  106.440002  104.738998  106.214996   

      Adj Close    Volume  
243  107.709999  20926300  
244  105.980003  20343100  
245  106.120003  17116300  
246  105.209999  19780600  
247  106.214996  20705300

## Descriptive Statistics
Descriptive Statistics will provide summary statistics for each stock in the dataset. 

In [6]:
# Descriptive Statistics for each stock
descriptive_stats = stocks_data.groupby('Ticker')['Close'].describe()
descriptive_stats

count        mean        std         min         25%         50%  \
Ticker                                                                     
AAPL     62.0  158.240645   7.360485  145.309998  152.077499  158.055000   
GOOG     62.0  100.631532   6.279464   89.349998   94.702501  102.759998   
MSFT     62.0  275.039839  17.676231  246.270004  258.742500  275.810013   
NFLX     62.0  327.614677  18.554419  292.760010  315.672493  325.600006   

               75%         max  
Ticker                          
AAPL    165.162506  173.570007  
GOOG    105.962503  109.459999  
MSFT    287.217506  310.649994  
NFLX    338.899994  366.829987

##  Time Series Analysis

In [7]:
stocks_data['Date'] = pd.to_datetime(stocks_data['Date'])
pivot_data = stocks_data.pivot(index='Date', columns='Ticker', values='Close')
pivot_data

Ticker            AAPL        GOOG        MSFT        NFLX
Date                                                      
2023-02-07  154.649994  108.040001  267.559998  362.950012
2023-02-08  151.919998  100.000000  266.730011  366.829987
2023-02-09  150.869995   95.459999  263.619995  362.500000
2023-02-10  151.009995   94.860001  263.100006  347.359985
2023-02-13  153.850006   95.000000  271.320007  358.570007
...                ...         ...         ...         ...
2023-05-01  169.589996  107.709999  305.559998  324.119995
2023-05-02  168.539993  105.980003  305.410004  317.549988
2023-05-03  167.449997  106.120003  304.399994  319.299988
2023-05-04  165.789993  105.209999  305.410004  320.779999
2023-05-05  173.570007  106.214996  310.649994  322.760010

[62 rows x 4 columns]

In [8]:
# Create a subplot
fig = make_subplots(rows=1, cols=1)

# Add traces for each stock ticker
for column in pivot_data.columns:
    fig.add_trace(
        go.Scatter(x=pivot_data.index, y=pivot_data[column], name=column),
        row=1, col=1
    )

# Update layout
fig.update_layout(
    title_text='Time Series of Closing Prices',
    xaxis_title='Date',
    yaxis_title='Closing Price',
    legend_title='Ticker',
    showlegend=True
)

fig

- The above plot displays the time series of the closing prices for each stock (AAPL, GOOG, MSFT, NFLX) over the observed period
- AAPL and MSFT exhibit a general upward trend in this period.
- NFLX shows more pronounced fluctuations compared to others.
- MSFT and NFLX generally trade at higher price levels than AAPL and GOOG in this dataset.

## Volatility Analysis
Volatility analysis refers to the examination of the degree of variation of a trading price series over time.

In [42]:
#  We’ll calculate and compare the volatility (standard deviation) of the closing prices for each stock.

volatility = pivot_data.std().sort_values(ascending=False)
volatility

Ticker
NFLX    18.554419
MSFT    17.676231
AAPL     7.360485
GOOG     6.279464
dtype: float64

In [45]:
fig = px.bar(volatility,
             x=volatility.index,
             y=volatility.values,
             labels={'y': 'Standard Deviation', 'x': 'Ticker'},
             title='Volatility of Closing Prices (Standard Deviation)')

fig

It indicates that NFLX and MSFT stocks were more prone to price fluctuations during this period compared to AAPL and GOOG.

## Correlation Analysis
Correlation analysis is a statistical method used to evaluate the strength and direction of the linear relationship between two or more variables.<br>
r = Correlation Coefficient

- r=1 indicates a perfect positive correlation.
- r=−1 indicates a perfect negative correlation.
- r=0 indicates no linear correlation.

In [49]:
# Calculate the correlation matrix
correlation_matrix = pivot_data.corr()
correlation_matrix

Ticker      AAPL      GOOG      MSFT      NFLX
Ticker                                        
AAPL    1.000000  0.901662  0.953037  0.154418
GOOG    0.901662  1.000000  0.884527  0.201046
MSFT    0.953037  0.884527  1.000000  0.191273
NFLX    0.154418  0.201046  0.191273  1.000000

In [52]:
fig = go.Figure(data=go.Heatmap(
                    z=correlation_matrix,
                    x=correlation_matrix.columns,
                    y=correlation_matrix.columns,
                    colorscale='blues',
                    colorbar=dict(title='Correlation'),
                    ))

# Update layout
fig.update_layout(
    title='Correlation Matrix of Closing Prices',
    xaxis_title='Ticker',
    yaxis_title='Ticker'
)

# Show the figure
fig.show()

- we can observe that there are varying degrees of positive correlations between the stock prices, with some pairs showing stronger correlations than others
- AAPL and MSFT seem to have a relatively higher positive correlation.

## Comparative Analysis
Comparative analysis in the stock market involves evaluating and comparing various aspects of different stocks or investment opportunities to make informed investment decisions

-  To compare the performance of different stocks based on their returns over the period.
-  To calculate the percentage change in closing prices from the start to the end of the period for each stock.

In [53]:
# Calculating the percentage change in closing prices
percentage_change = ((pivot_data.iloc[-1] - pivot_data.iloc[0]) / pivot_data.iloc[0]) * 100
percentage_change

Ticker
AAPL    12.234086
GOOG    -1.689193
MSFT    16.104798
NFLX   -11.073151
dtype: float64

In [54]:
fig = px.bar(percentage_change,
             x=percentage_change.index,
             y=percentage_change.values,
             labels={'y': 'Percentage Change (%)', 'x': 'Ticker'},
             title='Percentage Change in Closing Prices')

fig

- MSFT had highest positive change of approximately 16.10%.
- AAPL had exhibited a positive change of approximately 12.23%. It indicates a solid performance, though slightly lower than MSFT’s.
- GOOG had showed a slight negative change of about -1.69%. It indicates a minor decline in its stock price over the observed period.
- NFLX had experienced the most significant negative change, at approximately -11.07%. It suggests a notable decrease in its stock price during the period.

## Daily Risk Vs. Return Analysis

- Daily risk vs. Return analysis involves examining the relationship between the daily returns and associated risks (volatility) of individual stocks, portfolios, or indices.
- we will calculate the average daily return and the standard deviation of daily returns for each stock.

In [55]:
# The standard deviation will serve as a proxy for risk
# while the average daily return represents the expected return.

daily_returns = pivot_data.pct_change().dropna()
#  the pct_change() method in Pandas can be used to calculate the percentage change between the current and a prior element.
#  dropna() is used to remove the rows with missing values resulting from the percentage change calculation
daily_returns

Ticker          AAPL      GOOG      MSFT      NFLX
Date                                              
2023-02-08 -0.017653 -0.074417 -0.003102  0.010690
2023-02-09 -0.006912 -0.045400 -0.011660 -0.011804
2023-02-10  0.000928 -0.006285 -0.001972 -0.041766
2023-02-13  0.018807  0.001476  0.031243  0.032272
2023-02-14 -0.004225 -0.000526  0.003133  0.003876
...              ...       ...       ...       ...
2023-05-01 -0.000530 -0.004713 -0.005533 -0.017610
2023-05-02 -0.006191 -0.016062 -0.000491 -0.020270
2023-05-03 -0.006467  0.001321 -0.003307  0.005511
2023-05-04 -0.009913 -0.008575  0.003318  0.004635
2023-05-05  0.046927  0.009552  0.017157  0.006172

[61 rows x 4 columns]

In [57]:
# Recalculating average daily return and standard deviation (risk)
avg_daily_return = daily_returns.mean()
risk = daily_returns.std()
risk

Ticker
AAPL    0.014152
GOOG    0.020710
MSFT    0.017881
NFLX    0.022481
dtype: float64

In [58]:
# Creating a DataFrame for plotting
risk_return_df = pd.DataFrame({'Risk': risk, 'Average Daily Return': avg_daily_return})
risk_return_df

Risk  Average Daily Return
Ticker                                
AAPL    0.014152              0.001991
GOOG    0.020710             -0.000067
MSFT    0.017881              0.002606
NFLX    0.022481             -0.001676

In [61]:
fig = go.Figure()

# Add scatter plot points
fig.add_trace(go.Scatter(
    x=risk_return_df['Risk'],
    y=risk_return_df['Average Daily Return'],
    mode='markers+text',
    text=risk_return_df.index,
    textposition="top center",
    marker=dict(size=10)
))

# Update layout
fig.update_layout(
    title='Risk vs. Return Analysis',
    xaxis_title='Risk (Standard Deviation)',
    yaxis_title='Average Daily Return',
    showlegend=False
)

fig

- AAPL shows the lowest risk combined with a positive average daily return, suggesting a more stable investment with consistent returns.
- GOOG has a slightly negative daily return, indicating a riskier and less rewarding investment during this period.
- MSFT shows moderate risk with the highest average daily return, suggesting a potentially more rewarding investment, although with higher volatility compared to AAPL.
- NFLX exhibits the highest risk and a negative average daily return, indicating it was the most volatile and least rewarding investment among these stocks over the analyzed period.
